File: src/main/java/io/gitlab/hsedjame/buzz/data/dto/Messages.java
package io.gitlab.hsedjame.buzz.data.dto;

import java.util.List;

public sealed interface Messages {

    record Question(int number, String label, int points, List<Answer> answers) implements Messages{}

    record Answer(int number, String label, boolean good) implements Comparable<Answer>, Messages{
        @Override
        public int compareTo(Answer o) {
            return label.compareTo(o.label);
        }
    }

    record PlayerAnswer(String playerName, Answer answer) implements Messages{}

    record Buzz(String author) implements Messages{}

    record CanBuzz(boolean canBuzz) implements Messages {}

    record PlayerScore(String playerName, int score, String goodAnswer, boolean update) implements Messages{}

    record Error(String message) implements Messages {}
}

File: src/main/resources/templates/messages.js
export class PlayerScoreMsg {
     playerName;
     score;
     goodAnswer;

     constructor(name) {
         this.playerName = name;
         this.score = 0;
     }
}

export class BuzzMsg {
    author;
    time;
}

export class PlayerAnswerMsg {
    playerName;
    answer;
}

export class AnswerMsg {
    number;
    label;
    good;
}

export class Question {
    number;
    label;
    points;
    answers;
}

export class StateChange {
    type;
    canBuzz;
    message;
    players;
    requiredNbPlayers;
}

export const StateChangeType = Object.freeze(
    {
        "GAME_START" : "GAME_START",
        "GAME_END" : "GAME_END",
        "CAN_BUZZ" : "CAN_BUZZ",
        "NEW_PLAYER_SCORE" : "NEW_PLAYER_SCORE",
        "NEW_QUESTION" : "NEW_QUESTION",
        "NEW_BUZZ" : "NEW_BUZZ",
        "NEW_ANSWER" : "NEW_ANSWER"
    }
)